### PR TITLE
Themes: Fix back button on sheet

### PR DESCRIPTION
**Pull request by seear**
screen shot 2016-05-18 at 12 25 40

Back button will:
* Always go to /design...
* Maintain any filters or search args that were previously in place
* go to /design/ if arrived cold at /theme/...

I tried the approach of having the back path as a URL query arg, but it felt fragile. This approach stores the path in the global app state.

**To Test**
Try using the _< BACK_ button in the top left of the theme sheet in the following scenarios. Expected results as above.
* start from http://calypso.localhost:3000/design and choose details for a theme
* click around the different sections of the sheet
* activate a theme
* add a search at http://calypso.localhost:3000/design
* Use a filter (free/premium/all) at http://calypso.localhost:3000/design
* come in cold to /theme/{theme}
* logged-in
* logged-out

### CODE REVIEW

File: client/my-sites/theme/main.jsx [hunk 1]
```diff
@@ -127,12 +129,13 @@ const ThemeSheet = React.createClass( {
...
 
 		const { siteSlug, id } = this.props;
+		const sitePart = siteSlug ? `/${ siteSlug }` : '';
 
 		const nav = (
 			<NavTabs label="Details" >
 				{ this.getValidSections().map( ( section ) => (
 					<NavItem key={ section }
-						path={ `/theme/${ id }/${ section }/${ siteSlug }` }
+						path={ `/theme/${ id }/${ section }${ sitePart }` }
```
Comment (seear) [inline client/my-sites/theme/main.jsx]: the / on the end of the path here (in cases with no site) was causing an unnecessary redirect.

**Comment by folletto** [comment]
Added a commit with "All Themes" as the back text, as suggested by @kathrynwp in #5471.

screen shot 2016-05-20 at 22 36 07

**Comment by folletto** [comment]
Tested on Chrome, Firefox:

* Cold, logged out ✓
* Cold, logged in ✓ 
* From `/design` with no filter/search ✓
* From `/design` with free filter ✓
* From `/design` with search ✓
* From `/design` with premium filter + search ✓
* From `/design` with premium filter + search, using "Setup" ✓
* From `/design` with premium filter + search, using "Setup", logged out ✓

All work. 👍

**Comment by ockham** [comment]
> Added a commit with "All Themes" as the back text

If we're keeping filters and query string, it's not necessarily _All Themes_ tho, right? :-) What about something like _Themes Overview_?

**Comment by ockham** [comment]
Instead of introducing a new `themes-ui` Redux subtree -- could we instead re-use `themes-last-query`? AFAICS, its `getParams()` selector yields all the same information, so ideally we don't duplicate that...

**Comment by ockham** [comment]
The theme sheet's `main`'s `page` import is obsolete per 4b5167743703ee0a205e3d6fe5bad801a2f6e27a.

**Comment by seear** [comment]
> could we instead re-use themes-last-query? 

Whoa, great idea, let me look into that...

**Comment by folletto** [comment]
> If we're keeping filters and query string, it's not necessarily All Themes tho, right? :-) What about something like Themes Overview?

You're right, it's not strictly "all themes" but it's still the "all themes" page (with a filter), like "My Sites" might show just one site. "Overview" doesn't seem accurate either... but I agree we should change it to something better if we can find it. :)

In the future we can probably switch it depending on context:

* If it's a direct load or from outside .com, we show "All Themes"
* If it's from a specific section inside .com, we go back there (i.e. "Customize", "Front end")
* If it's from an unfiltered "Themes", then "All Themes".
* If it's from a filtered "Themes", then maybe just "Themes" or "Back to Themes" or "More Themes".

**Comment by seear** [comment]
> could we instead re-use themes-last-query? 

This works ok, see the gist:
https://gist.github.com/seear/8c50dcce0577e65e5c156772267b95cd

It has a couple of disadvantages over current technique in this PR:
* It encodes a lot of knowledge about our url structure into the selector as it builds the path up from the various params. This seems worse to me than a small bit of duplication, and will break when we get to magic search.
* It won't work when coming from anywhere other than /design (in which case there is actually no duplication)

Current technique seems more straightforward to me.

**Comment by ockham** [comment]
Okay, I think those are two pretty valid point. Let's stick with the new subtree then.

**Comment by ockham** [comment]
Rebase, squash, and :shipit:

**Pull request merged**